### ISSUE TITLE
Pickle files are 5x the size of text input.

### ISSUE THREAD

**Issue by antspy** [OpenNMT/OpenNMT-py#198]
Hi, 

I have downloaded the WMT2013 dataset (DE-EN translation), I have tokenized it and created the train set and test set. Altogether these files occupy around 600Mb of hard disk space. 

Then I pre-process them using onmt.IO.ONMTDataset, but the process is extremely slow and results in a 3.3 Gb train file. Even loading the train and test set from memory takes 8 minutes! Is this expected or am I doing something wrong?

Thank you!

**Comment by srush** [comment]
We'll look into this. Can you send the exact options that you used? It does pickle the training data, but I don't see why it would be that big. I'll give it a try on my side.

**Comment by antspy** [comment]
Thank you!

Note that I am working on the source code directly, but I essentially copied the preprocess.py file. This is what I do:
```
            print('Processing files ...')
            fields = onmt.IO.ONMTDataset.get_fields(self.trainFilesPath[0], self.trainFilesPath[1])
            print("Building Training...")
            train = onmt.IO.ONMTDataset(self.trainFilesPath[0], self.trainFilesPath[1], fields, opt)
            print("Building Vocab...")
            onmt.IO.ONMTDataset.build_vocab(train, opt)

            print("Building Test...")
            test = onmt.IO.ONMTDataset(self.testFilesPath[0], self.testFilesPath[1], fields, opt)
            print("Saving train/test/fields")
            with open(self.processedFilesPath[0], 'wb') as processFieldsFile:
                torch.save(fields, processFieldsFile, pickle_module=dill)
            with open(self.processedFilesPath[1], 'wb') as processedTrainFile:
                torch.save(train, processedTrainFile, pickle_module=dill)
            with open(self.processedFilesPath[2], 'wb') as processTestFile:
                torch.save(test, processTestFile, pickle_module=dill)
            print('Saving done')
```

where opt is just a named tuple with the following options (I took the default options from the file opts.py)

```
            opt = namedtuple('opt', 'src_seq_length tgt_seq_length tgt_seq_length_trunc dynamic_dict ' 
                    'src_seq_length_trunc src_vocab_size src_words_min_frequency tgt_vocab_size '
                    'tgt_words_min_frequency share_vocab')
            opt = opt(src_seq_length=50, tgt_seq_length=50, src_seq_length_trunc=0, tgt_seq_length_trunc=0,
                        dynamic_dict=True, src_vocab_size=50000, tgt_vocab_size=50000, src_words_min_frequency=0,
                        tgt_words_min_frequency=0, share_vocab=False)
```

Thank you!

**Comment by srush** [comment]
Can you try on the latest code? This week we eliminated the `dill` dependency, which I think should speed things up.

**Comment by srush** [comment]
I see the issue, yes we are seeing a 5x blowup in size. Will try to fix.

**Comment by vince62s** [comment]
not sure it is the same issue but I guess so.
preprocess.py 2 files that are 1.5GB and 1.3GB (EN DE too)
RAM usage topped at 98GB, taking for ever at saving time and crashed.

**Comment by playma** [comment]
@vince62s  I face the same problem. Have you solved it?

**Comment by srush** [comment]
So sharding is now checked in. Going to call this issue fixed for now. We still may need to do more work on the training side to use these files.
